service/history/workflow/timer_sequence.go

```diff
@@ -153,15 +153,14 @@ func (t *timerSequenceImpl) CreateNextActivityTimer() (bool, error) {
 	if !ok {
 		return false, serviceerror.NewInternal(fmt.Sprintf("unable to load activity info %v", firstTimerTask.EventID))
 	}
-
-	err := t.mutableState.UpdateActivity(activityInfo.ScheduledEventId, func(ai *persistencespb.ActivityInfo, ms MutableState) error {
-		// mark timer task mask as indication that timer task is generated
-		ai.TimerTaskStatus |= timerTypeToTimerMask(firstTimerTask.TimerType)
-		if firstTimerTask.TimerType == enumspb.TIMEOUT_TYPE_HEARTBEAT {
-			t.mutableState.UpdateActivityTimerHeartbeat(ai.ScheduledEventId, firstTimerTask.Timestamp)
-		}
-		return nil
-	})
+	// mark timer task mask as indication that timer task is generated
+	activityInfo.TimerTaskStatus |= timerTypeToTimerMask(firstTimerTask.TimerType)
+	var err error
+	if firstTimerTask.TimerType == enumspb.TIMEOUT_TYPE_HEARTBEAT {
+		err = t.mutableState.UpdateActivityTaskStatusWithTimerHeartbeat(activityInfo.ScheduledEventId, &activityInfo.TimerTaskStatus, &firstTimerTask.Timestamp)
+	} else {
+		err = t.mutableState.UpdateActivityTaskStatusWithTimerHeartbeat(activityInfo.ScheduledEventId, &activityInfo.TimerTaskStatus, nil)
+	}
```

Comment:
nit: just change the value of heartbeat time in the `if` and call UpdateActivityTaskStatusWithTimerHeartbeat once after the `if`?